### PR TITLE
[helioseasycontrols] Internationalization + actions for messages - Attempt 2

### PR DESCRIPTION
**Pull request by bern77**
This is attempt number 2 for the original PR #9704.

- Internationalization of UI relevant texts
- Better support for the error/warning/info/status messages received from the device (can now be retrieved via actions)

This PR already includes all changes committed into #9704.

Signed-off-by: Bernhard Bauer bern77@gmail.com

### CODE REVIEW

**Comment by bern77** [comment]
@fwolter: many thanks for your review of the previous PR. I see the following open points from #9704:

- Is `Map<String, Object>` the right return type for actions?
- Should the `configurationPid` of `HeliosEasyControlsTranslationProvider` be renamed (from `localization.helioseasycontrols` to e.g. `internationalization.helioseasycontrols`?

Also, you mentioned:
> There are also some checkstyle warnings left. You could take a look at target/code-analysis/report.html.
Where exactly do I find this report?

**Comment by fwolter** [comment]
> > There are also some checkstyle warnings left. You could take a look at target/code-analysis/report.html.
> 
> Where exactly do I find this report?

If you execute `mvn install` in your bundle's directory, there will appear the mentioned `target` directory.

I don't have an answer to the `configurationPid` topic currently.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsTranslationProvider.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * This class provides translated texts
+ *
+ * @author Bernhard Bauer - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = HeliosEasyControlsTranslationProvider.class, configurationPid = "localization.helioseasycontrols")
```
Comment (fwolter) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsTranslationProvider.java:33]: The configurationPid can be removed at all.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsHandler.java [hunk 2]
```diff
@@ -413,39 +425,44 @@ public void writeValue(String variableName, String value) throws HeliosException
...
+                                                lock.release();
+                                                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                                        "Error writing to device: "
+                                                                + failureInfo.getCause().getMessage());
+                                            });
+                                }
+                            } catch (InterruptedException e) {
+                                logger.warn(
+                                        "{} encountered Exception when trying to lock Semaphore for writing variable {} to the device: {}",
+                                        HeliosEasyControlsHandler.class.getSimpleName(), variableName, e.getMessage());
```
Comment (fwolter) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsHandler.java:459]: This exception is thrown when OH is shut-down. That means the current operation should be finished as fast as possible and no further operations should be done. You can achieve this by not catching the exception here but throw the InterruptedException and catch it at the highest level you can. So, this is not an error if this exception is raised.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsTranslationProvider.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+public class HeliosEasyControlsTranslationProvider {
+
+    private final Bundle bundle;
+    private final TranslationProvider i18nProvider;
+    private final LocaleProvider localeProvider;
+
+    @Activate
+    public HeliosEasyControlsTranslationProvider(@Reference TranslationProvider i18nProvider,
+            @Reference LocaleProvider localeProvider) {
+        this.bundle = FrameworkUtil.getBundle(this.getClass());
```
Comment (fwolter) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsTranslationProvider.java:43]: You could inject the BundleContext.
```suggestion
    public HeliosEasyControlsTranslationProvider(@Reference TranslationProvider i18nProvider,
            @Reference LocaleProvider localeProvider, BundleContext context) {
        this.bundle = context.getBundle();
```

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsHandler.java (hunk 2 above)
Comment (bern77) [reply bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsHandler.java]: Thanks for the input! I hope the changes I made are what you had in mind.

File: bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java [hunk 4]
```diff
@@ -54,10 +54,15 @@ public void setThingHandler(@Nullable ThingHandler handler) {
...
             if (handler != null) {
+
                 handler.writeValue(variableName, "1");
             }
         } catch (HeliosException e) {
             logger.warn("Error executing action triggering switch for variable {}: {}", variableName, e.getMessage());
+        } catch (InterruptedException e) {
+            logger.warn(
+                    "{} encountered Exception when trying to lock Semaphore for writing variable {} to the device: {}",
+                    HeliosEasyControlsHandler.class.getSimpleName(), variableName, e.getMessage());
```
Comment (fwolter) [inline bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsActions.java:65]: Your changes look good! As a shut-down of OH is a normal operation, you could make this debug or trace. It's even ok to not log it at all, but simply finish the program flow as your do it already.

**Review by fwolter — APPROVED** [review]
LGTM

**Pull request merged**